src/Stratis.Bitcoin.Features.Consensus/ProvenBlockHeaders/ProvenBlockHeaderStore.cs

```diff
@@ -0,0 +1,363 @@
+        /// <param name="asyncLoopFactory">Factory for creating and also possibly starting application defined tasks inside async loop.</param>
+        public ProvenBlockHeaderStore(
+            ConcurrentChain chain,
+            IDateTimeProvider dateTimeProvider,
+            ILoggerFactory loggerFactory,
+            IProvenBlockHeaderRepository provenBlockHeaderRepository,
+            INodeLifetime nodeLifetime,
+            INodeStats nodeStats,
+            IAsyncLoopFactory asyncLoopFactory)
+        {
+            Guard.NotNull(chain, nameof(chain));
+            Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            Guard.NotNull(provenBlockHeaderRepository, nameof(provenBlockHeaderRepository));
+            Guard.NotNull(nodeLifetime, nameof(nodeLifetime));
+            Guard.NotNull(nodeStats, nameof(nodeStats));
+            Guard.NotNull(asyncLoopFactory, nameof(asyncLoopFactory));
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.chain = chain;
+            this.provenBlockHeaderRepository = provenBlockHeaderRepository;
+            this.nodeLifetime = nodeLifetime;
+
+            this.PendingBatch = new ConcurrentDictionary<int, ProvenBlockHeader>();
+            this.Cache = new ConcurrentDictionary<int, ProvenBlockHeader>();
+
+            this.asyncLoopFactory = asyncLoopFactory;
+
+            this.performanceCounter = new BackendPerformanceCounter(dateTimeProvider);
+            nodeStats.RegisterStats(this.AddBenchStats, StatsType.Benchmark, 400);
+        }
+
+        /// <inheritdoc />
+        public async Task InitializeAsync()
+        {
+            await this.provenBlockHeaderRepository.InitializeAsync().ConfigureAwait(false);
+
+            this.TipHashHeight = this.provenBlockHeaderRepository.TipHashHeight;
+
+            this.logger.LogDebug("Initialized ProvenBlockHader block tip at '{0}'.", this.TipHashHeight);
+
+            this.asyncLoop = this.asyncLoopFactory.Run("ProvenBlockHeaders job", token =>
+            {
+                // Save pending items.
+                this.SaveAsync().ConfigureAwait(false);
+
+                // Check and make sure the cache limit hasn't been breached.
+                this.ManangeCacheSize();
+
+                this.logger.LogTrace("(-)[IN_ASYNC_LOOP]");
+
+                return Task.CompletedTask;
+            },
+            this.nodeLifetime.ApplicationStopping,
+            repeatEvery: TimeSpan.FromMinutes(1),
+            startAfter: TimeSpan.FromMinutes(1));
+        }
+
+        /// <inheritdoc />
+        public async Task<ProvenBlockHeader> GetAsync(int blockHeight)
+        {
+            ProvenBlockHeader header = null;
+
+            using (new StopwatchDisposable(o => this.performanceCounter.AddQueryTime(o)))
+            {
+                header = this.CheckCache(blockHeight);
+
+                if (header == null)
+                {
+                    // Check the repository (DBreeze).
```

Comment:
thats implementation detail. Repository might be not on top of dbreeze, so I'd not specify the db name